core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -196,7 +208,7 @@ private boolean doEmit() {
     }
 
     private void handleCancelState() {
-        synchronized (this) {
+        synchronized (cancelLock) {
```

Comment:
What if `!onStreanCalled().get()` is no longer true after we make the check?

Reply:
`onStreanCalled` will only go from `false -> true` once and never back to false.

Reply:
Aaah I see what you mean. Looking at it, yep, it might cause a NPE at [line 293](https://github.com/aws/aws-sdk-java-v2/pull/5176/files/059fa2a71a7a1e4d5bbb5eec6523bbc27fc1257b#diff-b9ec282c15a46057d63a0021282180d584a26dc1c229a66e8a95c3c706dfa87dR293). Adding synchronization around the `if (onStreamCalled.compareAndSet(false, true))` on `cancelLock` there at line 293 should prevent it.